### PR TITLE
Refactor Middleware to Use Strategy Pattern Across Backends

### PR DESCRIPTION
**Pull request by Theo3883**
# Pull Request: Refactor Middleware to Use Strategy Pattern Across Backends

## Overview
This PR introduces a significant refactor to the authentication and authorization middleware in all backend services (`backend-AI`, `backend-portfolio`, `backend-user`). The core change is the adoption of the **Strategy Pattern** for the OAuth2 middleware, resulting in improved modularity, testability, and maintainability. Extensive new tests and abstractions have been added to support this architecture.

---

### 1. Strategy Pattern Implementation
- **New Interfaces Introduced:**
  - `IAuthenticationStrategy`: Defines `CanHandle(HttpContext)` and `AuthenticateAsync(HttpContext)` for pluggable authentication logic.
  - `IAuthenticationContextService`: Orchestrates multiple strategies, selecting the first that can handle the request.
  - `IAuthorizationPathService`, `ISecurityHeadersService`, and (in backend-user) `IClaimsBuilderService`: Encapsulate related concerns for authorization and security.
- **Concrete Implementations:**
  - `OAuth2AuthenticationStrategy`: Handles Bearer token authentication, supports case-insensitive "Bearer", trims whitespace, and logs outcomes.
  - `AuthenticationContextService`: Accepts a collection of strategies and delegates authentication to the first capable one.
  - `AuthorizationPathService`, `SecurityHeadersService`, `ClaimsBuilderService`: Provide modular, testable logic for their respective concerns.
- **Dependency Injection:**
  - All new services and strategies are registered in DI via `Program.cs` or `ServiceCollectionExtensions.cs`.
  - Extension methods like `AddAuthenticationServices()` and `AddCustomAuthenticationStrategy<T>()` allow easy registration and chaining.

---

### 2. Middleware Refactor
- **`OAuth2Middleware`** in all backends:
  - No longer contains direct authentication logic.
  - Now depends on `IAuthenticationContextService` and other abstractions.
  - Handles public endpoints, missing/invalid headers, and sets user claims on success.
  - Returns 401 for all authentication failures, with clear logging.

---

### 3. Testing Improvements
- **Unit Tests for Each Service:**
  - `OAuth2AuthenticationStrategyTests`: Verifies correct handling of Bearer tokens, case insensitivity, whitespace, invalid schemes, and error handling.
  - `AuthenticationContextServiceTests`: Ensures correct strategy selection, null handling, and error cases.
  - `ServiceCollectionExtensionsTests`: Verifies correct DI registration, lifetimes, and chaining for custom strategies.
  - Middleware tests (`OAuth2MiddlewareTests`): Cover all edge cases—public endpoints, missing/invalid headers, invalid/valid tokens, and user claim setting.

### CODE REVIEW

File: backend/backend-user/Services/Abstractions/IAuthenticationContextService.cs [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+    /// Follows Strategy pattern as the context that manages authentication strategies.
+    /// </summary>
+    public interface IAuthenticationContextService
+    {
+        /// <summary>
+        /// Attempts to authenticate the request using available strategies.
+        /// </summary>
+        /// <param name="context">The HTTP context to authenticate.</param>
+        /// <returns>The authenticated user if successful, null otherwise.</returns>
+        Task<User?> AuthenticateAsync(HttpContext context);
```
Comment (RazvanIonita03) [inline backend/backend-user/Services/Abstractions/IAuthenticationContextService.cs:16]: Same issue as the interface IAuthenticationStrategy, inconsistent return types for the same interface in multiple microservices. Should be changed to match the others.

**Comment by Theo3883** [comment]
### Summary
- Unified authentication interfaces in `backend-user` to return ClaimsPrincipal, matching `backend-portfolio` and `backend-AI`.
- Refactored strategy/context/middleware to operate on ClaimsPrincipal directly.
- Updated tests and verified all backends still pass.

### What changed
- Interfaces:
  - `backend/backend-user/Services/Abstractions/IAuthenticationStrategy.cs`: return type → `Task<ClaimsPrincipal?>`
  - `backend/backend-user/Services/Abstractions/IAuthenticationContextService.cs`: return type → `Task<ClaimsPrincipal?>`
- Implementations:
  - `backend/backend-user/Services/OAuth2AuthenticationStrategy.cs`: builds and returns `ClaimsPrincipal?` after token validation
  - `backend/backend-user/Services/AuthenticationContextService.cs`: orchestrates strategies returning `ClaimsPrincipal?`
  - `backend/backend-user/Middleware/OAuth2Middleware.cs`: uses the principal from the auth context; no local claims-building
- Tests updated to reflect new contracts:
  - `backend/backend-user.tests/Services/AuthenticationContextServiceTests.cs`
  - `backend/backend-user.tests/Services/OAuth2AuthenticationStrategyTests.cs`
  - `backend/backend-user.tests/Middleware/OAuth2MiddlewareTests.cs`
  - `backend/backend-user.tests/Extensions/ServiceCollectionExtensionsTests.cs`

### Dependency injection pattern
- All microservices, including `backend-user`, already use extension-based registration via `AddApplicationServices()` in `Config/ServiceRegistrationConfiguration.cs` and call it from `Program.cs`.

### Backward compatibility / impact
- Internal breaking change within `backend-user` interfaces only; all references updated.
- Middleware now receives a `ClaimsPrincipal` directly; `IClaimsBuilderService` remains available but is no longer invoked in middleware.

File: backend/backend-AI/Services/Abstractions/IAuthenticationStrategy.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+using System.Security.Claims;
+using Microsoft.AspNetCore.Http;
+
+namespace backend_AI.Services.Abstractions
+{
+    /// <summary>
+    /// Strategy interface for different authentication mechanisms.
+    /// </summary>
+    public interface IAuthenticationStrategy
```
Comment (RazvanIonita03) [inline backend/backend-AI/Services/Abstractions/IAuthenticationStrategy.cs:9]: This interface does a bit too much. Consider splitting this into smaller, focused interfaces.
